File: kotlin/provenance-mark/src/test/kotlin/com/blockchaincommons/provenancemark/DateTest.kt
package com.blockchaincommons.provenancemark

import com.blockchaincommons.dcbor.CborDate
import java.time.Instant
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertFails

class DateTest {
    @Test
    fun test2ByteDates() {
        val baseDate = CborDate.fromYmdHms(2023, 6, 20, 0, 0, 0)
        val serialized = baseDate.serialize2Bytes()
        assertEquals("00d4", serialized.toHex())
        val deserialized = deserialize2Bytes(serialized)
        assertEquals(baseDate, deserialized)

        val minSerialized = byteArrayOf(0x00, 0x21)
        val minDate = CborDate.fromYmdHms(2023, 1, 1, 0, 0, 0)
        val deserializedMin = deserialize2Bytes(minSerialized)
        assertEquals(minDate, deserializedMin)

        val maxSerialized = byteArrayOf(0xff.toByte(), 0x9f.toByte())
        val deserializedMax = deserialize2Bytes(maxSerialized)
        val expectedMaxDate = CborDate.fromYmdHms(2150, 12, 31, 0, 0, 0)
        assertEquals(expectedMaxDate, deserializedMax)

        val invalidSerialized = byteArrayOf(0x00, 0x5e)
        assertFails { deserialize2Bytes(invalidSerialized) }
    }

    @Test
    fun test4ByteDates() {
        val baseDate = CborDate.fromYmdHms(2023, 6, 20, 12, 34, 56)
        val serialized = baseDate.serialize4Bytes()
        assertContentEquals(hex("2a41d470"), serialized)
        val deserialized = deserialize4Bytes(serialized)
        assertEquals(baseDate, deserialized)

        val minSerialized = hex("00000000")
        val minDate = CborDate.fromYmdHms(2001, 1, 1, 0, 0, 0)
        val deserializedMin = deserialize4Bytes(minSerialized)
        assertEquals(minDate, deserializedMin)

        val maxSerialized = hex("ffffffff")
        val deserializedMax = deserialize4Bytes(maxSerialized)
        val expectedMaxDate = CborDate.fromYmdHms(2137, 2, 7, 6, 28, 15)
        assertEquals(expectedMaxDate, deserializedMax)
    }

    @Test
    fun test6ByteDates() {
        val baseDate = CborDate.fromInstant(Instant.parse("2023-06-20T12:34:56.789Z"))
        val serialized = baseDate.serialize6Bytes()
        assertContentEquals(hex("00a51125d895"), serialized)
        val deserialized = deserialize6Bytes(serialized)
        assertEquals(baseDate, deserialized)

        val minSerialized = hex("000000000000")
        val minDate = CborDate.fromYmdHms(2001, 1, 1, 0, 0, 0)
        val deserializedMin = deserialize6Bytes(minSerialized)
        assertEquals(minDate, deserializedMin)

        val maxSerialized = hex("e5940a78a7ff")
        val deserializedMax = deserialize6Bytes(maxSerialized)
        val expectedMaxDate = CborDate.fromInstant(Instant.parse("9999-12-31T23:59:59.999Z"))
        assertEquals(expectedMaxDate, deserializedMax)

        val invalidSerialized = hex("e5940a78a800")
        assertFails { deserialize6Bytes(invalidSerialized) }
    }
}

File: typescript/provenance-mark/tests/date.test.ts
import { describe, test, expect } from 'vitest';
import { bytesToHex, CborDate } from '@bc/dcbor';
import {
  serialize2Bytes, deserialize2Bytes,
  serialize4Bytes, deserialize4Bytes,
  serialize6Bytes, deserialize6Bytes,
} from '../src/index.js';

describe('date serialization', () => {
  test('2-byte dates', () => {
    const baseDate = CborDate.fromYmd(2023, 6, 20);
    const serialized = serialize2Bytes(baseDate);
    expect(bytesToHex(serialized)).toBe('00d4');
    const deserialized = deserialize2Bytes(serialized);
    expect(baseDate.equals(deserialized)).toBe(true);

    // Minimum date
    const minSerialized = new Uint8Array([0x00, 0x21]);
    const minDate = CborDate.fromYmd(2023, 1, 1);
    const deserializedMin = deserialize2Bytes(minSerialized);
    expect(minDate.equals(deserializedMin)).toBe(true);

    // Maximum date
    const maxSerialized = new Uint8Array([0xff, 0x9f]);
    const deserializedMax = deserialize2Bytes(maxSerialized);
    const expectedMaxDate = CborDate.fromYmd(2150, 12, 31);
    expect(deserializedMax.equals(expectedMaxDate)).toBe(true);

    // Invalid date (2023-02-30)
    const invalidSerialized = new Uint8Array([0x00, 0x5e]);
    expect(() => deserialize2Bytes(invalidSerialized)).toThrow();
  });

  test('4-byte dates', () => {
    const baseDate = CborDate.fromYmdHms(2023, 6, 20, 12, 34, 56);
    const serialized = serialize4Bytes(baseDate);
    expect(bytesToHex(serialized)).toBe('2a41d470');
    const deserialized = deserialize4Bytes(serialized);
    expect(baseDate.equals(deserialized)).toBe(true);

    // Minimum date
    const minSerialized = new Uint8Array([0x00, 0x00, 0x00, 0x00]);
    const minDate = CborDate.fromYmd(2001, 1, 1);
    const deserializedMin = deserialize4Bytes(minSerialized);
    expect(minDate.equals(deserializedMin)).toBe(true);

    // Maximum date
    const maxSerialized = new Uint8Array([0xff, 0xff, 0xff, 0xff]);
    const deserializedMax = deserialize4Bytes(maxSerialized);
    const expectedMaxDate = CborDate.fromYmdHms(2137, 2, 7, 6, 28, 15);
    expect(deserializedMax.equals(expectedMaxDate)).toBe(true);
  });

  test('6-byte dates', () => {
    // Base date with milliseconds
    const dt = new Date(Date.UTC(2023, 5, 20, 12, 34, 56, 789));
    const baseDate = CborDate.fromDatetime(dt);
    const serialized = serialize6Bytes(baseDate);
    expect(bytesToHex(serialized)).toBe('00a51125d895');
    const deserialized = deserialize6Bytes(serialized);
    expect(baseDate.equals(deserialized)).toBe(true);

    // Minimum date
    const minSerialized = new Uint8Array([0x00, 0x00, 0x00, 0x00, 0x00, 0x00]);
    const minDate = CborDate.fromYmd(2001, 1, 1);
    const deserializedMin = deserialize6Bytes(minSerialized);
    expect(minDate.equals(deserializedMin)).toBe(true);

    // Maximum date
    const maxSerialized = new Uint8Array([0xe5, 0x94, 0x0a, 0x78, 0xa7, 0xff]);
    const deserializedMax = deserialize6Bytes(maxSerialized);
    const expectedDt = new Date(Date.UTC(9999, 11, 31, 23, 59, 59, 999));
    const expectedMaxDate = CborDate.fromDatetime(expectedDt);
    expect(deserializedMax.equals(expectedMaxDate)).toBe(true);

    // Invalid date (exceeds maximum)
    const invalidSerialized = new Uint8Array([0xe5, 0x94, 0x0a, 0x78, 0xa8, 0x00]);
    expect(() => deserialize6Bytes(invalidSerialized)).toThrow();
  });
});
